_posts/2021-01-08-consumer-tuning.md

```diff
@@ -0,0 +1,280 @@
+
+![Consumer lag](/assets/images/posts/2021-01-08-consumer-lag.png)
+
+### Improving throughput by increasing the minimum amount of data fetched in a request
+
+Use the `fetch.max.wait.ms` and `fetch.min.bytes` configuration properties to set thresholds that control the number of requests from your consumer.
+
+* **`fetch.max.wait.ms`**
+Sets a maximum threshold for time-based batching.
+
+* **`fetch.min.bytes`**
+Sets a minimum threshold for size-based batching.
+
+When the client application polls for data, both these properties govern the amount of data fetched by the consumer from the broker. You can adjust the properties higher so that there are fewer requests, and messages are delivered in bigger batches. You might want to do this if the amount of data being produced is low. Reducing the number of requests from the consumer also lowers the overhead on CPU utilization for the consumer and broker.
+
+Throughput is improved by increasing the amount of data fetched in a batch, but with some cost to latency. Consequently, adjusting these properties lower has the effect of lowering end-to-end latency. We'll look a bit more at targeting latency by increasing batch sizes in the next section.
+
+You can use one or both or these properties. If you use both, Kafka will respond to a fetch request when the first of either threshold is reached.
+
+```properties
+# ...
+fetch.max.wait.ms=500
+fetch.min.bytes=16384
+# ...
+```
+
+## Lowering latency by increasing the maximum batch sizes
+
+Increasing the minimum amount of data fetched in a request can help with increasing throughput. But if you want to do something to improve latency, you can extend your thresholds by increasing the _maximum_ amount of data that can be fetched by the consumer from the broker.
+
+* **`fetch.max.bytes`**
+Sets a maximum limit in bytes on the amount of data fetched from the broker at one time
+
+* **`max.partition.fetch.bytes`**
+Sets a maximum limit in bytes on how much data is returned for each partition, which must always be larger than the number of bytes set in the broker or topic configuration for `max.message.bytes`.
+
+By increasing the values of these two properties, and allowing more data in each request, latency is improved as there are fewer fetch requests.
+
+```properties
+# ...
+fetch.max.bytes=52428800
+max.partition.fetch.bytes=1048576
+# ...
+```
+
+> Be careful with memory usage. The maximum amount of memory a client can consume is calculated approximately as:
+>
+> `_NUMBER-OF-BROKERS_ * fetch.max.bytes and _NUMBER-OF-PARTITIONS_ * max.partition.fetch.bytes`
```

Comment:
Correct markup for italics?

Reply:
Bit of formatting required with `pre` tags.

> Be careful with memory usage. The maximum amount of memory a client can consume is calculated approximately as:
>
> <pre><i>NUMBER-OF-BROKERS</i> * fetch.max.bytes and <i>NUMBER-OF-PARTITIONS</i> * max.partition.fetch.bytes</pre>